src/plugins/discord/fetchDiscord.js

```diff
@@ -0,0 +1,198 @@
+// @flow
+
+import retry from "retry";
+import {type Emoji, type BotToken, type Snowflake} from "./models";
+import {DiscordMirror} from "./mirror";
+import {type endCursor, type pageInfo, Fetcher} from "./fetcher";
+
+const DISCORD_SERVER = "https://discordapp.com/api";
+
+type FetchPlanOptions = {|
+  lastMessageId: string,
+  lastMemberId: string,
+  lastReactionId: string,
+|};
+
+/**
+ * Fetches provided url wwith provided options.
+ * Returns either the response.json or rejects with
+ * a retry flag and an optional message.
+ *
+ */
+function tryDiscordFetch(fetch, url, fetchOptions): Promise<any> {
+  return fetch(url, fetchOptions).then((response) => {
+    switch (response.status) {
+      case 403:
+        return Promise.reject({
+          retry: false,
+          message: `403 Forbidden: bad API username or key?\n${response.url}`,
+        });
+      case 404:
+        return Promise.reject({
+          retry: false,
+          message: `404 Not Found on: ${response.url}; maybe bad serverUrl?`,
+        });
+      case 410:
+        return Promise.reject({retry: false, message: `410 Gone`});
+      case 429:
+        return Promise.reject({retry: true, message: null});
+      default:
+        return Promise.resolve(response.json());
+    }
+  });
+}
+
+// If retry = true in tryDiscordFetch reject, will re-attempt the fetch.
+function retryDiscordFetch(fetch, url, fetchOptions) {
+  return new Promise((resolve, reject) => {
+    const operation = retry.operation();
```

Comment:
Noticed I didn't mention why randomized is important. Randomizing is very useful in avoiding deadlocks in distributed / parallel cases. And since we're already distributed + parallel simply from running multiple plugins at once, this should be useful in all cases :]

Reply:
This fell out of the PR scope, resolving.